app/src/main/java/com/github/steroidteam/todolist/view/NoteSelectionFragment.java

```diff
@@ -35,43 +39,119 @@ public View onCreateView(
         recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
         recyclerView.setHasFixedSize(true);
 
-        notes = new ArrayList<>();
-        NoteAdapter.NoteCustomListener customListener = this::openNote;
-
-        adapter = new NoteAdapter(notes, customListener);
+        adapter = new NoteArrayListAdapter(createCustomListener());
         recyclerView.setAdapter(adapter);
 
-        database = DatabaseFactory.getDb();
-        database.getNotesList()
-                .thenAccept(
-                        uuids -> {
-                            for (UUID uuid : uuids) {
-                                database.getNote(uuid)
-                                        .thenAccept(
-                                                note -> {
-                                                    notes.add(note);
-                                                    adapter.notifyDataSetChanged();
-                                                });
-                            }
+        viewModel = new NoteSelectionViewModel();
+        viewModel
+                .getNoteList()
+                .observe(
+                        getActivity(),
+                        (noteList) -> {
+                            adapter.setNoteList(noteList);
+                            adapter.notifyDataSetChanged();
                         });
 
         return root;
     }
 
-    public void createNote(View view) {
-        Note newNote = new Note("New note");
+    private NoteArrayListAdapter.NoteHolder.NoteCustomListener createCustomListener() {
+        return new NoteArrayListAdapter.NoteHolder.NoteCustomListener() {
+            @Override
+            public void onClickCustom(NoteArrayListAdapter.NoteHolder holder) {
+                Bundle bundle = new Bundle();
+                bundle.putSerializable(NOTE_ID_KEY, holder.getId());
+                Navigation.findNavController(holder.itemView)
+                        .navigate(R.id.nav_note_display, bundle);
+            }
 
-        database.putNote(newNote.getId(), newNote)
-                .thenAccept(
-                        str -> {
-                            notes.add(newNote);
-                            adapter.notifyDataSetChanged();
-                        });
+            @Override
+            public void onNoteDelete(NoteArrayListAdapter.NoteHolder holder) {
+                removeNote(holder.getId());
+            }
+
+            @Override
+            public void onLongClickCustom(Note note) {
+                renameNote(note);
+            }
+        };
+    }
+
+    private void createNote(View view) {
```

Comment:
I think it needs some factorization with removeNote and renameNote. They have a lot of duplicate code